concepts/type-theory-glossary.md

```diff
@@ -0,0 +1,396 @@
+
+### enums
+
+Enums have a cardinality that is some integer.
+
+For example, an enum might have members named "Foo", "Bar", and "Baz";
+and the cardinality of that enum is "|3|".
+
+(The term "enum" might make more sense if you think of it as short for
+*enumerated*, rather than *enumerable* -- something with countable-infinity
+members is **not** considered an "enum".
+(We'd call that an [infinite scalar](#infinite-scalar) instead.))
+
+### sum types
+
+Sum types have a cardinality that is *the sum of cardinality of their member types*.
+
+Sum types usually show up in prose descriptions of data as the word "or".
+
+For example, say: "I have a boolean *or* an int" --
+this would be an example of a sum type,
+and its cardinality would be "|inf+2|" --
+countable infinity for the int, and plus two for the boolean.
+
+A clearer example might take two enums -- let's call them
+"Animals" (it has members "cat", "dog", "rabbit"),
+and "Furniture" (it has members "chair", and "desk") --
+and say "I have an Animal *or* a furniture" --
+then clearly I have one (and only one) of these five things:
+the cardinality is "|3+2|".
+
+Sum types go by many names in various communities: they're also commonly
+known as "(discriminated) unions", "(tagged) variants", or other terminology.
+(We use the "union" term ourselves, in the lexicon of IPLD Schemas!)
+
+You can also consult the Wikipedia page on
+[Sum Types](https://en.wikipedia.org/wiki/Sum_types)
+for broader discussion of the topic.
+https://chadaustin.me/2015/07/sum-types/ also contains some useful content
+and examples from a variety of languages.
+
+### product types
+
+Product types have a cardinality that is *the multiplied cardinality of their member types*.
+
+This may sound outlandish, but you've been using them all your life:
+*"product types" are commonly known as "structs"*
+(or, "classes", in more aggressively object-oriented languages).
+
+Product types usually show up in prose descriptions of data as the word "and".
+
+For example, say: "I have a boolean and another boolean and another boolean" --
+you might write code for this as `struct{ a bool, b bool, c bool }`,
+and its cardinality would be "|2*2*2|" -- "|8|" --
+two for each of the booleans, and multiplied because each of them may vary independently,
+and thus each of the booleans adds another "dimension" to the state space described by this combined value.
+
+For another example, using our enums again above --
+"Animals" (it has members "cat", "dog", "rabbit"),
+and "Furniture" (it has members "chair", and "desk") --
+we could say "I have an Animal, *and* it's on top of a Furniture" --
+and then I could have any one of of *six* states
+("cat on a chair", "cat on desk", "dog on a chair", etc):
+the cardinality is "|3*2|".
+
+### infinite recursives
+
+Infinite recursives, as it says in the name, have infinite cardinality.
+
+Maps and lists are both examples of infinite recursives.
```

Comment:
By "recursive," I was assuming you meant a type that can contain itself, but this sort of depends on what's meant by "map" and "list" here. In general, the statement you're making is true of "containers", whether or not they're recursive. I suppose all fp-style containers can be represented as recursive types, so maybe it doesn't matter? (Is there another category of types that represents the general concept of a "container" as used in e.g. C++?)

Reply:
Maybe just add a clarification, something like:

>  Maps and lists are both examples of infinite recursives in that they can contain a theoretically unbounded number of additional maps and lists.

This terminology has come up in the schema documentation and it's worked quite well there I think.

Reply:
Made another run at this riffing from @rvagg's text.